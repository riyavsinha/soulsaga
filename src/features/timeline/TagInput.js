import CancelIcon from '@material-ui/icons/Cancel';
import Chip from '@material-ui/core/Chip';
import Downshift from 'downshift';
import MenuItem from '@material-ui/core/MenuItem';
import Paper from '@material-ui/core/Paper';
import PropTypes from 'prop-types';
import TextField from '@material-ui/core/TextField';
import React, { Component } from 'react';
import deburr from 'lodash/deburr';

export default class TagInput extends Component {
  static propTypes = {
    availableTags: PropTypes.arrayOf(PropTypes.string).isRequired,
    eventTags: PropTypes.arrayOf(PropTypes.string).isRequired,
    onChange: PropTypes.func.isRequired,
  };

  state = { inputValue: '' };

  /**
   * If comma entered, treats input as a completed tag and adds, otherwise
   * updates input state normally.
   *
   * @param {!Event} e 'change' event
   */
  handleInputChange = e => {
    const input = e.target.value;
    if (input.slice(-1) === ',') {
      this.addTag_(input.slice(0, input.length - 1));
    } else {
      this.setState({ inputValue: e.target.value });
    }
  };

  /**
   * Listens for 'Enter' key clicks, and adds current input as tag
   * if suggestion menu is not open, or there is no highlighted value
   * in the open menu.
   *
   * @param {boolean} isOpen Whether the menu is open
   * @param {?number} index The index of the selected item in the menu
   */
  handleInputEnter = (isOpen, index) => e => {
    if ((!isOpen || index === null) && e.key === 'Enter') {
      this.addTag_(e.target.value);
    }
  };

  /**
   * For non-empty tags not already added to the list, adds the tag
   * to the list and alerts parent using onChange. Also clears input
   * field.
   *
   * @param {string} tag
   * @private
   */
  addTag_ = tag => {
    if (!this.props.eventTags.includes(tag) && tag.length !== 0) {
      this.props.eventTags.push(tag);
      this.props.onChange(this.props.eventTags);
    }
    this.setState({ inputValue: '' });
  };

  /**
   * Handles changes initiated by Downshift, when a tag suggestion
   * is selected.
   *
   * @param {string} tag
   */
  handleChange = tag => this.addTag_(tag);

  /**
   * Handles deletion of an an added tag when 'x' is clicked in Chip.
   *
   * @param {string} item The tag item to delete
   */
  handleDelete = item => () => {
    const tags = this.props.eventTags;
    tags.splice(tags.indexOf(item), 1);
    this.props.onChange(tags);
  };

  /**
   * Generates suggestions for Downshift menu. Suggestion is applicable if
   * input value is a substring of the suggestion.
   *
   * Operates on normalized version of tags so tags with special characters
   * can still be found without having to re-enter them.
   *
   * @param {string} value The input value to find suggestions for.
   */
  getSuggestions = value => {
    const normalizeStr = s => deburr(s.trim()).toLowerCase();
    const inputValue = normalizeStr(value);

    return this.props.availableTags.filter(item =>
      normalizeStr(item).includes(inputValue),
    );
  };

  render() {
    let layoutClass =
      this.props.mode === 'side' ? 'timeline-tag-input__flex' : '';
    return (
      <div className="timeline-tag-input">
        <Downshift
          onChange={this.handleChange}
          selectedItem={this.props.eventTags}
          inputValue={this.state.inputValue}
        >
          {({
            getInputProps,
            getItemProps,
            isOpen,
            toggleMenu,
            inputValue,
            selectedItem,
            highlightedIndex,
          }) => (
            <div className={layoutClass}>
              <TagList
                selectedItems={selectedItem}
                onDelete={this.handleDelete}
              />

              <TextField
                value={inputValue}
                label="Tags"
                className="timeline-tag-input__text-field"
                InputProps={getInputProps({
                  onChange: this.handleInputChange,
                  onKeyDown: this.handleInputEnter(isOpen, highlightedIndex),
                  onClick: () => toggleMenu(),
                })}
              />

              <SuggestionMenu
                suggestions={this.getSuggestions(inputValue)}
                isOpen={isOpen}
                getItemProps={getItemProps}
                highlightedIndex={highlightedIndex}
                selectedItem={selectedItem}
              />
            </div>
          )}
        </Downshift>
      </div>
    );
  }
}

/** A list of tag chips */
function TagList({ selectedItems, onDelete }) {
  if (!selectedItems.length) {
    return null;
  }
  return (
    <div className="timeline-tag-input__tag-chip-section">
      {selectedItems.map(item => (
        <Chip
          key={item + 'Chip'}
          label={item}
          onDelete={onDelete(item)}
          className="timeline-tag-input__tag-chip"
        />
      ))}
    </div>
  );
}
TagList.propTypes = {
  onDelete: PropTypes.func.isRequired,
  selectedItems: PropTypes.arrayOf(PropTypes.string).isRequired,
};

/** A tag suggestion menu item in Downshift */
function Suggestion({
  suggestion,
  index,
  itemProps,
  highlightedIndex,
  selectedItem,
}) {
  const isHighlighted = highlightedIndex === index;
  const isSelected = selectedItem.indexOf(suggestion.label) > -1;
  return (
    <MenuItem
      {...itemProps}
      selected={isHighlighted}
      component="div"
      style={{
        fontWeight: isSelected ? 500 : 400,
      }}
    >
      {suggestion}
    </MenuItem>
  );
}
Suggestion.propTypes = {
  highlightedIndex: PropTypes.number,
  index: PropTypes.number,
  itemProps: PropTypes.object,
  selectedItem: PropTypes.arrayOf(PropTypes.string).isRequired,
  suggestion: PropTypes.string.isRequired,
};

function SuggestionMenu({
  suggestions,
  isOpen,
  getItemProps,
  highlightedIndex,
  selectedItem,
}) {
  if (!isOpen || !suggestions.length) {
    return null;
  }
  return (
    <Paper square className="timeline-tag-input__autocomplete-menu">
      {suggestions.map((suggestion, index) => (
        <Suggestion
          key={suggestion + 'Option'}
          suggestion={suggestion}
          index={index}
          itemProps={getItemProps({ item: suggestion })}
          highlightedIndex={highlightedIndex}
          selectedItem={selectedItem}
        />
      ))}
    </Paper>
  );
}
SuggestionMenu.propTypes = {
  suggestions: PropTypes.arrayOf(PropTypes.string).isRequired,
  isOpen: PropTypes.bool.isRequired,
  getItemProps: PropTypes.func.isRequired,
  highlightedIndex: PropTypes.number,
  selectedItem: PropTypes.arrayOf(PropTypes.string).isRequired
}