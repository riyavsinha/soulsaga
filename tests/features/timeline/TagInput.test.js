import Downshift from 'downshift';
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import TagInput from 'features/timeline/TagInput';

const MENU_CLASS = '.timeline-tag-input__autocomplete-menu';

describe('timeline/TagInput', () => {
  let props;
  let onChangeFn;

  beforeEach(() => {
    onChangeFn = jest.fn();
    props = {
      availableTags: [],
      eventTags: [],
      onChange: onChangeFn,
    };
  });

  describe('snapshot tests', () => {
    it('renders empty', () => {
      const renderedComponent = renderer
        .create(<TagInput {...props} />)
        .toJSON();

      expect(renderedComponent).toMatchSnapshot();
    });

    it('renders with tags and available tags, available tags hidden', () => {
      props.availableTags = ['testAvailableTag1', 'testAvailableTag2'];
      props.eventTags = ['testEventTag1', 'testEventTag2'];
      const renderedComponent = renderer
        .create(<TagInput {...props} />)
        .toJSON();

      expect(renderedComponent).toMatchSnapshot();
    });
  });

  describe('Chip list tests', () => {
    it('renders number of chips in eventTags', () => {
      props.eventTags = ['testEventTag1', 'testEventTag2'];

      const wrapper = mount(<TagInput {...props} />);

      expect(wrapper.find('Chip')).toHaveLength(2);
    })

    it('deletes tag when chip delete icon clicked', () => {
      props.eventTags = ['testEventTag1', 'testEventTag2'];

      const wrapper = mount(<TagInput {...props} />);

      expect(wrapper.find('Chip').first().find('svg')).toHaveLength(1);
      wrapper.find('Chip').first().find('svg').simulate('click');

      expect(onChangeFn).toHaveBeenCalledTimes(1);
      expect(onChangeFn).toHaveBeenLastCalledWith(['testEventTag2']);
    })
  })

  describe('Input tests', () => {
    it('enters text on normal letter presses', () => {
      const wrapper = mount(<TagInput {...props} />);

      wrapper.find('input').simulate('change', { target: { value: 'ea' } });

      expect(wrapper.state().inputValue).toBe('ea');
      expect(onChangeFn).not.toHaveBeenCalled();
    });

    it('clears input and adds unique tag when comma entered', () => {
      const wrapper = mount(<TagInput {...props} />);

      wrapper.find('input').simulate('change', { target: { value: 'ea,' } });

      expect(wrapper.state().inputValue).toBe('');
      expect(onChangeFn).toHaveBeenCalledTimes(1);
      expect(onChangeFn).toHaveBeenLastCalledWith(['ea']);
    });

    it('appends new tag to existing tags', () => {
      props.eventTags = ['work', 'beach']
      const wrapper = mount(<TagInput {...props} />);

      wrapper.find('input').simulate('change', { target: { value: 'food,' } });

      expect(onChangeFn).toHaveBeenCalledTimes(1);
      expect(onChangeFn).toHaveBeenLastCalledWith(['work', 'beach', 'food']);
    });

    it('clears input when comma entered, non-unique tag not added', () => {
      props.eventTags = ['ea'];
      const wrapper = mount(<TagInput {...props} />);

      wrapper.find('input').simulate('change', { target: { value: 'ea,' } });

      expect(wrapper.state().inputValue).toBe('');
      expect(onChangeFn).not.toHaveBeenCalled();
    });

    it('adds tag when enter clicked', () => {
      const wrapper = mount(<TagInput {...props} />);

      wrapper.find('input').simulate('change', { target: { value: 'ea' } });
      wrapper.find('input').simulate('keydown', { key: 'Enter' });

      expect(wrapper.state().inputValue).toBe('');
      expect(onChangeFn).toHaveBeenCalledTimes(1);
      expect(onChangeFn).toHaveBeenLastCalledWith(['ea']);
    });
  });

  describe('Menu tests', () => {
    it('toggles suggestion menu on input click', () => {
      props.availableTags = ['eat']
      
      const wrapper = mount(<TagInput {...props} />);
      expect(wrapper.find(MENU_CLASS).length).toBeFalsy();

      wrapper.find('input').simulate('click');
      expect(wrapper.find(MENU_CLASS).length).toBeTruthy();

      wrapper.find('input').simulate('click');
      expect(wrapper.find(MENU_CLASS).length).toBeFalsy();
    });

    it('does not open when no suggestions available', () => {      
      const wrapper = mount(<TagInput {...props} />);
      expect(wrapper.find(MENU_CLASS).length).toBeFalsy();

      wrapper.find('input').simulate('click');
      expect(wrapper.find(MENU_CLASS).length).toBeFalsy();
    });

    it('renders suggestions that input value is subset of when strings normalized', () => {
      props.availableTags = ['eat', 'béach', 'flEa', 'new york', 'work'];

      const wrapper = mount(<TagInput {...props} />);
      wrapper.find('input').simulate('change', {target: {value: 'ea'}});

      const menuItemsWrapper = wrapper.find('MenuItem');
      expect(menuItemsWrapper).toHaveLength(3);
      expect(menuItemsWrapper.at(0).text()).toBe('eat');
      expect(menuItemsWrapper.at(1).text()).toBe('béach');
      expect(menuItemsWrapper.at(2).text()).toBe('flEa');
    });

    it('adds menu item to tags when clicked', () => {
      props.availableTags = ['work'];

      const wrapper = mount(<TagInput {...props} />);
      wrapper.find('input').simulate('click');

      const menuItemsWrapper = wrapper.find('MenuItem').simulate('click');

      expect(onChangeFn).toHaveBeenCalledTimes(1);
      expect(onChangeFn).toHaveBeenLastCalledWith(['work']);
    });
  });
});
