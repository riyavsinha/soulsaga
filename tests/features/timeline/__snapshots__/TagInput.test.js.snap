// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline/TagInput snapshot tests renders empty 1`] = `
<div
  className="timeline-tag-input"
>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    aria-owns={null}
    className=""
    role="combobox"
  >
    <div
      className="MuiFormControl-root-1 timeline-tag-input__text-field"
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink={false}
      >
        Tags
      </label>
      <div
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-controls={null}
        aria-labelledby="downshift-0-label"
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          className="MuiInputBase-input-46 MuiInput-input-31"
          disabled={false}
          id="downshift-0-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`timeline/TagInput snapshot tests renders with tags and available tags, available tags hidden 1`] = `
<div
  className="timeline-tag-input"
>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-labelledby="downshift-1-label"
    aria-owns={null}
    className=""
    role="combobox"
  >
    <div
      className="timeline-tag-input__tag-chip-section"
    >
      <div
        className="MuiChip-root-53 MuiChip-deletable-59 timeline-tag-input__tag-chip"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          className="MuiChip-label-72"
        >
          testEventTag1
        </span>
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-78 MuiChip-deleteIcon-73"
          focusable="false"
          onClick={[Function]}
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
      </div>
      <div
        className="MuiChip-root-53 MuiChip-deletable-59 timeline-tag-input__tag-chip"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          className="MuiChip-label-72"
        >
          testEventTag2
        </span>
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-78 MuiChip-deleteIcon-73"
          focusable="false"
          onClick={[Function]}
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 timeline-tag-input__text-field"
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink={false}
      >
        Tags
      </label>
      <div
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-controls={null}
        aria-labelledby="downshift-1-label"
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          className="MuiInputBase-input-46 MuiInput-input-31"
          disabled={false}
          id="downshift-1-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
